### PR TITLE
Add an inspect icon for node element when possible

### PR DESCRIPTION
**Pull request by nchevobbe**
The icon should be displayed if a nodeActor matching the grip is passed as a prop

### CODE REVIEW

**Comment by nchevobbe** [comment]
✅✅✅✅✅✅✅✅ :)

Now that everything is green, asking @juliandescottes for review on this.

I'm not sure about the name of  the `attachedNodeFront` props. I wanted to convey the sense that the nodeFront should be in the DOM tree to show the icon, but maybe we could find a better name.

Also, I'm not sure if we should compute `attachedNodeFront` from `attachedNodeFrontByActor` in each "container" component. Maybe we could only pass `attachedNodeFrontByActor` to element-node and text-node component and let them check the presence of a nodeFront that matches the grip.

Please tell me what do you think on this, thanks !

File: src/reps/rep-utils.js [hunk 1]
```diff
@@ -157,6 +160,66 @@ function wrapRender(renderMethod) {
...
+
+function getPreviewGrips(grip, filterFn = null) {
+  let grips = [];
+  if (!filterFn || filterFn(grip)) {
+    grips.push(grip);
+  }
+
+  let previewItems = [];
+  if (grip && grip.preview) {
+    if (grip.preview.items) {
```
Comment (juliandescottes) [inline src/reps/rep-utils.js]: Some comments would be nice in this section, to know which block corresponds to which kind of grip.

File: src/reps/rep-utils.js [hunk 2]
```diff
@@ -157,6 +160,66 @@ function wrapRender(renderMethod) {
...
+  if (!Array.isArray(parameters)) {
+    return null;
+  }
+  let nodeGrips = parameters.reduce((result, parameter) => {
+    return result.concat(getPreviewGrips(parameter, filterFn));
+  }, []);
+  return [...new Set(nodeGrips)];
+}
+
+function getPreviewGrips(grip, filterFn = null) {
```
Comment (juliandescottes) [inline src/reps/rep-utils.js]: Can we add a JS doc here?

File: src/reps/rep-utils.js [hunk 3]
```diff
@@ -167,5 +230,8 @@ module.exports = {
   parseURLParams,
   parseURLEncodedText,
   getFileName,
-  getURLDisplayString
+  getURLDisplayString,
+  isGripSelectableInInspector,
+  getFlattenedGrips,
```
Comment (juliandescottes) [inline src/reps/rep-utils.js]: Is getFlattenedGrips used outside of this file? If not I wouldn't export it as it's pretty specific.

File: src/reps/rep-utils.js [hunk 4]
```diff
@@ -157,6 +160,66 @@ function wrapRender(renderMethod) {
...
+}
+
+function getFlattenedGrips(parameters, filterFn) {
+  if (!Array.isArray(parameters)) {
+    return null;
+  }
+  let nodeGrips = parameters.reduce((result, parameter) => {
+    return result.concat(getPreviewGrips(parameter, filterFn));
+  }, []);
+  return [...new Set(nodeGrips)];
```
Comment (juliandescottes) [inline src/reps/rep-utils.js]: Needs a comment explaining why we want unique values here.

File: src/reps/rep-utils.js (hunk 3 above)
Comment (juliandescottes) [inline src/reps/rep-utils.js]: I can't actually find where these new methods are being used ?
Is this supposed to be an external API used by some tools not in this repo 
(hey I wouldn't have to ask this in a mono repo setup :) )

File: src/reps/event.js [hunk 5]
```diff
@@ -16,7 +18,14 @@ let Event = React.createClass({
...
   propTypes: {
-    object: React.PropTypes.object.isRequired
+    object: React.PropTypes.object.isRequired,
+    objectLink: React.PropTypes.func,
+    // @TODO Change this to Object.values once it's supported in Node's version of V8
+    mode: React.PropTypes.oneOf(Object.keys(MODE).map(key => MODE[key])),
+    attachedNodeFrontsByActor: React.PropTypes.object,
+    onDOMNodeMouseOver: React.PropTypes.func,
+    onDOMNodeMouseOut: React.PropTypes.func,
+    onInspectIconClick: React.PropTypes.func,
```
Comment (juliandescottes) [inline src/reps/event.js]: Hmmm looks like I forgot to synchronize the eslint fixes from MC, will need to check if other files also need updating.

File: src/reps/promise.js [hunk 6]
```diff
@@ -41,13 +45,30 @@ const PromiseRep = React.createClass({
...
       return PropRep(Object.assign({}, this.props, {
         mode: MODE.TINY,
         name: `<${key}>`,
-        object: promiseState[key],
+        object,
         equal: ": ",
-        delim: i < keys.length - 1 ? ", " : ""
+        delim: i < keys.length - 1 ? ", " : "",
+        attachedNodeFrontsByActor,
+        attachedNodeFront: (
```
Comment (juliandescottes) [inline src/reps/promise.js]: attachedNodeFront is not part of the declared propTypes for PropRep and looking at the code of PropRep it doesn't seem to be used.

Comment (nchevobbe) [inline src/reps/promise.js]: true !

File: src/reps/rep-utils.js (hunk 3 above)
Comment (nchevobbe) [inline src/reps/rep-utils.js]: We do use functions in the console for this, I'll check if some aren't.

**Comment by nchevobbe** [comment]
> The overall approach makes sense to me. We already have the "objectLink" which is using a different pattern to allow to "customize" reps from the outside.
> Did you consider using the same approach here with an inspectorLink component ? Honestly not sure which one is best (except for consistency purposes).

That's a good point. I'll think of a way of doing this, it looks more consistent with the `objectLink`pattern.

> It would be nice to see some consumer code for the console to see how this fits.

Sure, I'll add some call points from the console in the next update of this PR

**Comment by nchevobbe** [comment]
@juliandescottes I simplified things a little and expanded the test to use the rep-utils function.
I think this should have the css and images to show the icon too, which I'm still working on and shouldn't impact what's in this PR so far.
Also, I will make the icon appears in the Reps tester by passing it a stubs for `attachedNodeFrontByActor` like below : 

```diff
diff --git a/src/launchpad/components/Result.js b/src/launchpad/components/Result.js
index dc7e9de..285bd81 100644
--- a/src/launchpad/components/Result.js
+++ b/src/launchpad/components/Result.js
@@ -4,6 +4,7 @@ const { DOM: dom, PropTypes, createFactory } = React;
 const { MODE } = require("../../reps/constants");
 const Rep = createFactory(require("../../reps/rep").Rep);
 const Grip = require("../../reps/grip");
+const { getSelectableInInspectorGrips } = require("../../reps/rep-utils");

 const Result = React.createClass({
   propTypes: {
@@ -47,7 +48,18 @@ const Result = React.createClass({
         key: JSON.stringify(object) + modeKey,
         "data-mode": modeKey,
       },
-      Rep({ object, defaultRep: Grip, mode: MODE[modeKey] })
+      Rep({
+        object,
+        defaultRep: Grip,
+        mode: MODE[modeKey],
+        onInspectIconClick: nodeFront => console.log("inspectIcon click", nodeFront),
+        attachedNodeFrontsByActor: getSelectableInInspectorGrips(object)
+          .reduce((res, gripStub) => {
+            return Object.assign(res, {
+              [gripStub.actor]: gripStub
+            });
+          }, {})
+      })
     );
   },
```

**Comment by nchevobbe** [comment]
Managed to have the icon in the Rep tester, I'll update the PR shortly :) 
screenshot at fevr 09 20-11-09

**Comment by nchevobbe** [comment]
@juliandescottes I think this is good for review now.
@jasonLaster would you mind having a look at https://github.com/devtools-html/devtools-reps/pull/52/commits/e347eafd20abdaa8d963ebc460f93dbc8e1be95c ? I took what was done in the debugger for Svg and I want to make sure I'm not doing anything stupid. 

Thank you both !

**Comment by nchevobbe** [comment]
Rebased on top of https://github.com/devtools-html/devtools-reps/commit/5b5f8e6482a45de8bea55bf7670a6d7dd1d16ea7

File: src/reps/grip-array.js [hunk 7]
```diff
@@ -161,15 +165,27 @@ let GripArrayItem = React.createFactory(React.createClass({
...
-
+    let {
+      attachedNodeFrontsByActor,
+    } = this.props;
     return (
       span({},
         Rep(Object.assign({}, this.props, {
-          mode: MODE.TINY
+          mode: MODE.TINY,
+          attachedNodeFrontsByActor,
```
Comment (juliandescottes) [inline src/reps/grip-array.js]: Not necessary since we are already using Object.assign on this.props?

File: src/reps/images/Svg.js [hunk 8]
```diff
@@ -0,0 +1,21 @@
+const React = require("react");
+const InlineSVG = require("svg-inline-react");
+
+const svg = {
+  "open-inspector": require("./open-inspector.svg"),
+};
+
+module.exports = function(name, props) { // eslint-disable-line
```
Comment (nchevobbe) [inline src/reps/images/Svg.js]: I copied it from debugger.html, let me check why it's needed

File: src/reps/grip.js [hunk 9]
```diff
@@ -133,7 +137,8 @@ const GripRep = React.createClass({
         object: value,
         equal: ": ",
         delim: i !== indexes.length - 1 || truncate ? ", " : "",
-        defaultRep: Grip
+        defaultRep: Grip,
+        attachedNodeFrontsByActor: this.props.attachedNodeFrontsByActor,
```
Comment (nchevobbe) [inline src/reps/grip.js]: right !

File: src/reps/promise.js [hunk 10]
```diff
@@ -41,13 +45,25 @@ const PromiseRep = React.createClass({
...
         name: `<${key}>`,
-        object: promiseState[key],
+        object,
         equal: ": ",
-        delim: i < keys.length - 1 ? ", " : ""
+        delim: i < keys.length - 1 ? ", " : "",
+        attachedNodeFrontsByActor,
+        onDOMNodeMouseOver,
+        onDOMNodeMouseOut,
+        onInspectIconClick,
```
Comment (nchevobbe) [inline src/reps/promise.js]: right !

File: src/reps/prop-rep.js [hunk 11]
```diff
@@ -29,36 +29,47 @@ let PropRep = React.createClass({
...
-      key = Rep({
-        object: this.props.name,
-        mode: this.props.mode || MODE.TINY,
+      key = Rep(Object.assign({}, this.props, {
+        object: name,
+        mode: mode || MODE.TINY,
         defaultRep: Grip,
-        objectLink: this.props.objectLink,
-      });
+        attachedNodeFrontsByActor,
```
Comment (nchevobbe) [inline src/reps/prop-rep.js]: 👍

**Comment by nchevobbe** [comment]
Thanks for the review @juliandescottes , I fixed the comments and pushed again, I'll merge if Circle is green

**Pull request merged**